### PR TITLE
[fix] complex type promotion

### LINKED ISSUE #76803: Complex x float type promotion is incorrect
Typical complex x float type promotion between tensors that are rank 1+ results in a complex result with the highest "value type" among the inputs. For example:

```
a = torch.randn((2, 2), dtype=torch.cfloat)
b = torch.randn((2, 2), dtype=torch.double)
(a + b).dtype
: torch.complex128
```

In the above the complex comes from the cfloat tensor, and the double from the double tensor. And that's fine.

When working with rank 0 x rank 1+ tensors, however, the behavior is incorrect:

```
a = torch.randn((2, 2), dtype=torch.cfloat)
b = torch.tensor(1, dtype=torch.double)
(a + b).dtype
: torch.complex64

a = torch.randn((2, 2), dtype=torch.float)
b = torch.tensor(1, dtype=torch.cdouble)
(a + b).dtype
: torch.complex128
```

In the first case the datatype of the tensor with rank 2 is used, and the higher "value type" of the rank 0 tensor is ignored. This is consistent with our type promotion behavior, which biases towards tensors of rank 1+.

In the second case, however, the "value type" of the rank 0 complex tensor is used. This seems like an error.

cc @ezyang @anjali411 @dylanbespalko @mruberry @Lezcano @nikitaved @nairbv @ngimel

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes https://github.com/pytorch/pytorch/issues/76803

Before Fix:
```python
>> a = torch.randn((2, 2), dtype=torch.float)
>> b = torch.tensor(1, dtype=torch.cdouble)
>> (a + b).dtype
torch.complex128
```

After Fix:
```python
>> a = torch.randn((2, 2), dtype=torch.float)
>> b = torch.tensor(1, dtype=torch.cdouble)
>> (a + b).dtype
torch.complex64
```

**Note**: This is a BC Breaking change.

### CODE REVIEW

File: aten/src/ATen/native/TypeProperties.cpp [hunk 1]
```diff
@@ -80,11 +80,12 @@ static inline ScalarType combine_categories(ScalarType higher, ScalarType lower)
   // NOLINTNEXTLINE(bugprone-branch-clone)
   if(isComplexType(higher)) {
```
Comment (kshitij12345) [inline aten/src/ATen/native/TypeProperties.cpp:81]: TODO: Add comment about the rule to combine categories

File: torch/_prims/utils.py [hunk 2]
```diff
@@ -823,32 +823,11 @@ def _find_highest_dtype_filtered(
             torch.get_default_dtype() if result_dtype is None else result_dtype
         )
     elif highest_type is complex:
-        # NOTE: complex x float type promotion is incorrectly implemented in PyTorch today
```
Comment (kshitij12345) [inline torch/_prims/utils.py:826]: cc: @mruberry

**Comment by kshitij12345** [comment]
ScalarXScalar is still broken. Will update once that is ready.

File: test/test_ops_gradients.py [hunk 3]
```diff
@@ -10,8 +10,9 @@
 from torch.testing._internal.common_device_type import \
     (instantiate_device_type_tests, ops, OpDTypes)
 
-# TODO: fixme https://github.com/pytorch/pytorch/issues/68972
-torch.set_default_dtype(torch.float32)
```
Comment (albanD) [inline test/test_ops_gradients.py:14]: Not sure about this change? We actually WANT to keep this as `float32` as it is the default!

**Review by albanD — COMMENTED** [review]
Not sure what is the exact problem with gradcheck. You should only ever get double precision numbers there anyways. (float64 or complex128).

File: test/test_ops_gradients.py (hunk 3 above)
Comment (kshitij12345) [reply test/test_ops_gradients.py:14]: (I agree that changing the default is not what we want ideally but just wanted to confirm. Sadly ScalarXScalar is still broken so can't get green signal from CI but changing the default dtype to `double` works locally)

The following script fails with this PR (based on failed test from `test_ops_gradient.py`)

```python
import torch

from torch.testing._internal.common_methods_invocations import op_db

# torch.set_default_dtype(torch.double) # Fails if this is not set.

op = list(filter(lambda op: (op.name == 'H'), op_db))[0]

samples = op.sample_inputs('cpu', torch.complex128, requires_grad=True)

for sample in samples:
    torch.autograd.gradgradcheck(op.op, sample.input, check_fwd_over_rev=True)
```

Haven't verified yet, but maybe the `gradcheck/gradgradcheck` code splits complex128 tensor into `real` (double tensor) and `imag` (double tensor) part does computation and tries to combine them back.

For default dtype of `torch.float`,
Before this PR, `1j * torch.double -> torch.complex128`, everything passed. (but wrong type-promotion).
After this PR, `1j * torch.double -> torch.complex64`, tests start failing. (correct type-promotion).

Does this explain the probable issue @albanD?

File: aten/src/ATen/native/TypeProperties.cpp [hunk 4]
```diff
@@ -80,11 +80,12 @@ static inline ScalarType combine_categories(ScalarType higher, ScalarType lower)
...
     return higher;
```
Comment (anjali411) [inline aten/src/ATen/native/TypeProperties.cpp:82]: which branch do we take for this case:
```py
a = torch.randn((2, 2), dtype=torch.float)
b = torch.tensor(1, dtype=torch.double)
print((a+b).dtype) #float
```

File: test/test_ops_gradients.py (hunk 3 above)
Comment (albanD) [reply test/test_ops_gradients.py:14]: > After this PR, 1j * torch.double -> torch.complex64, tests start failing. (correct type-promotion).

That is the right type promotion? That would be loosing precision for the double arg no?

**Comment by anjali411** [comment]
```py
>>> s = torch.tensor(1, dtype=torch.cfloat)
>>> t = torch.ones(3, dtype=torch.double)
>>> (s * t).dtype
torch.complex128
```
This seems like the correct behavior. 

I think the issue is this:
```py
>>> s = torch.tensor(1, dtype=torch.cfloat)
>>> t = torch.ones(3, dtype=torch.double)
>>> (s * t).dtype # expected cfloat
torch.complex128
```

File: aten/src/ATen/native/TypeProperties.cpp (hunk 4 above)
Comment (kshitij12345) [reply aten/src/ATen/native/TypeProperties.cpp:82]: For this case, we take

https://github.com/pytorch/pytorch/blob/e4cb9f8bf1ab0f41a3f75dba0ae012a693cc5182/aten/src/ATen/native/TypeProperties.cpp#L85-L86

**Comment by kshitij12345** [comment]
```python
import torch

patch_applied = True

# TensorXScalar
# HalfTensor X LongScalar -> Half
t = torch.ones(3, dtype=torch.half)
s = torch.ones((), dtype=torch.long)
# half (existing)
assert (t * s).dtype == torch.half

# LongTensor X HalfScalar -> Half
t = torch.ones(3, dtype=torch.long)
s = torch.ones((), dtype=torch.half)
# half (existing)
assert (t * s).dtype == torch.half

# DoubleTensor X CFloatScalar -> ?
t = torch.ones(3, dtype=torch.double)
s = torch.ones((), dtype=torch.cfloat)
if not patch_applied:
    # before PR: cdouble
    assert (t * s).dtype == torch.cdouble
else:
    # after PR: cfloat
    assert (t * s).dtype == torch.cfloat

# CHalfTensor X DoubleScalar -> CHalf
t = torch.ones(3, dtype=torch.chalf)
s = torch.ones((), dtype=torch.double)
# existing
assert (t * s).dtype == torch.chalf

# DoubleTensor X CHalfScalar -> ?
t = torch.ones(3, dtype=torch.double)
s = torch.ones((), dtype=torch.chalf)
if not patch_applied:
    # before PR: cdouble
    assert (t * s).dtype == torch.cdouble
else:
    # after PR: chalf
    assert (t * s).dtype == torch.chalf

# ScalarXScalar

# Half X Long -> Half
s1 = torch.ones((), dtype=torch.half)
s2 = torch.ones((), dtype=torch.long)
# half (existing)
assert (s1 * s2).dtype == torch.half

# CHalf X Double -> ?
s1 = torch.ones((), dtype=torch.chalf)
s2 = torch.ones((), dtype=torch.double)
if not patch_applied:
    # before PR: cdouble
    assert (t * s).dtype == torch.cdouble
else:
    # after PR: chalf
    assert (s1 * s2).dtype == torch.chalf

```
Does this make sense?

cc: @anjali411 @mruberry

File: test/test_ops_gradients.py (hunk 3 above)
Comment (mruberry) [reply test/test_ops_gradients.py:14]: a complex scalar multiplied with a cdouble tensor should produce a cdouble tensor, though. The type promotion for float x complex is different from long x float.

**Comment by mruberry** [comment]
> ```python
> import torch
> 
> patch_applied = True
> 
> # TensorXScalar
> # HalfTensor X LongScalar -> Half
> t = torch.ones(3, dtype=torch.half)
> s = torch.ones((), dtype=torch.long)
> # half (existing)
> assert (t * s).dtype == torch.half

This is correct. 

> # LongTensor X HalfScalar -> Half
> t = torch.ones(3, dtype=torch.long)
> s = torch.ones((), dtype=torch.half)
> # half (existing)
> assert (t * s).dtype == torch.half

This is correct.

> # DoubleTensor X CFloatScalar -> ?
> t = torch.ones(3, dtype=torch.double)
> s = torch.ones((), dtype=torch.cfloat)
> if not patch_applied:
>     # before PR: cdouble
>     assert (t * s).dtype == torch.cdouble
> else:
>     # after PR: cfloat
>     assert (t * s).dtype == torch.cfloat

This should be cdouble. because the tensor with 1+ dimensions sets the value type.

> # CHalfTensor X DoubleScalar -> CHalf
> t = torch.ones(3, dtype=torch.chalf)
> s = torch.ones((), dtype=torch.double)
> # existing
> assert (t * s).dtype == torch.chalf

This is correct.

> # DoubleTensor X CHalfScalar -> ?
> t = torch.ones(3, dtype=torch.double)
> s = torch.ones((), dtype=torch.chalf)
> if not patch_applied:
>     # before PR: cdouble
>     assert (t * s).dtype == torch.cdouble
> else:
>     # after PR: chalf
>     assert (t * s).dtype == torch.chalf

This should be cdouble.

> # ScalarXScalar
> 
> # Half X Long -> Half
> s1 = torch.ones((), dtype=torch.half)
> s2 = torch.ones((), dtype=torch.long)
> # half (existing)
> assert (s1 * s2).dtype == torch.half

This is correct.

> # CHalf X Double -> ?
> s1 = torch.ones((), dtype=torch.chalf)
> s2 = torch.ones((), dtype=torch.double)
> if not patch_applied:
>     # before PR: cdouble
>     assert (t * s).dtype == torch.cdouble
> else:
>     # after PR: chalf
>     assert (s1 * s2).dtype == torch.chalf
> ```

This should be cdouble.

File: torch/_prims/utils.py [hunk 5]
```diff
@@ -823,32 +823,11 @@ def _find_highest_dtype_filtered(
...
-            result_dtype = _find_highest_dtype_filtered(
-                args,
-                lambda x: is_float_dtype(x) or is_complex_dtype(x),
-                float_as_complex=True,
-                all_tensors_equal=True,
-            )
-
+        result_dtype = _find_highest_dtype_filtered(
+            args,
+            lambda x: is_complex_dtype(x),
```
Comment (mruberry) [inline torch/_prims/utils.py:828]: When the highest type is complex this needs to look at both float and complex tensors to determine the output dtype, so I think this lambda filter still needs to include float types

File: test/test_ops_gradients.py (hunk 3 above)
Comment (kshitij12345) [reply test/test_ops_gradients.py]: Thanks for the clarity! I think I understand it better now. Apologies for the noise :)!

File: aten/src/ATen/native/TypeProperties.cpp [hunk 6]
```diff
@@ -80,15 +80,28 @@ static inline ScalarType combine_categories(ScalarType higher, ScalarType lower)
...
-  }
-  else if(!isComplexType(lower) && isFloatingType(higher)) {
+  } else if (isComplexType(lower)) {
+    // special case for bfloat16
+    if (higher == kBFloat16) {
```
Comment (kshitij12345) [inline aten/src/ATen/native/TypeProperties.cpp:85]: `toComplexType` doesn't handle `BFloat16`.

File: aten/src/ATen/native/TypeProperties.cpp [hunk 7]
```diff
@@ -80,15 +80,28 @@ static inline ScalarType combine_categories(ScalarType higher, ScalarType lower)
...
+      // BFloat16 corresponds to Float in terms of range,
+      // so we return ComplexFloat
+      return kComplexFloat;
```
Comment (anjali411) [inline aten/src/ATen/native/TypeProperties.cpp:88]: what if lower if complexhalf? (I think the expected out dtype is complexhalf in that case)

File: aten/src/ATen/native/TypeProperties.cpp [hunk 8]
```diff
@@ -80,15 +80,28 @@ static inline ScalarType combine_categories(ScalarType higher, ScalarType lower)
...
+    }
+    // preserve value type of higher if it is floating type.
+    if (isFloatingType(higher)) {
+      return toComplexType(higher);
```
Comment (anjali411) [inline aten/src/ATen/native/TypeProperties.cpp:92]: same as above, if higher if float and lower is chalf, the expected out dtype is chalf

Comment (kshitij12345) [reply aten/src/ATen/native/TypeProperties.cpp:92]: If I understand correctly, value type of `higher` (1D+ tensor incase of 1D+ tensor and scalar inputs) should preserve the value type of `higher`.

Also to verify I checked against the type_promotion utility in `_prims/util.py`.

```python
import torch
from torch._prims.utils import elementwise_dtypes, ELEMENTWISE_TYPE_PROMOTION_KIND

s = torch.tensor(1j, dtype=torch.chalf)  # lower
t = torch.zeros(3, dtype=torch.float)  # higher

print(elementwise_dtypes(s, t, type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT))

```

Output
```
(torch.complex64, torch.complex64) # (comp_dtype, result_dtype)
```

File: aten/src/ATen/native/TypeProperties.cpp (hunk 7 above)
Comment (kshitij12345) [reply aten/src/ATen/native/TypeProperties.cpp:88]:
```python
import torch
from torch._prims.utils import elementwise_dtypes, ELEMENTWISE_TYPE_PROMOTION_KIND

s = torch.tensor(1j, dtype=torch.chalf)
t = torch.zeros(3, dtype=torch.bfloat16)

print(elementwise_dtypes(s, t, type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT))
```

Output
```
(torch.complex64, torch.complex64)
```

File: test/test_binary_ufuncs.py [hunk 9]
```diff
@@ -746,8 +746,9 @@ def _supported(dtypes):
             if _supported((torch.complex128,)):
                 rhs_c128_scalar_tensor = make_rhs_scalar_tensor(dtype=torch.complex128)
                 result = op(lhs_f32, rhs_c128_scalar_tensor)
+                # Value type of 1D+ Tensor (lhs_f32) takes priority over scalar tensor (rhs_c128).
                 expected_dtype = (
-                    torch.complex128 if not op.always_returns_bool else torch.bool
+                    torch.complex64 if not op.always_returns_bool else torch.bool
```
Comment (anjali411) [inline test/test_binary_ufuncs.py:751]: what is this change for?

**Review by anjali411 — APPROVED** [review]
Thanks for the type promotion fix! also, cool test improvements :D

**Comment by kshitij12345** [comment]
Unable to reproduce the CI failure locally. (Flaky?)

CI Ref: https://github.com/pytorch/pytorch/runs/6500788273?check_suite_focus=true#step:9:8642

<details>
```
======================================================================
[8642](https://github.com/pytorch/pytorch/runs/6500788273?check_suite_focus=true#step:9:8643)
FAIL [0.141s]: test_binary_op_scalarlist_fastpath__foreach_div_cuda_float16 (__main__.TestForeachCUDA)
[8643](https://github.com/pytorch/pytorch/runs/6500788273?check_suite_focus=true#step:9:8644)
----------------------------------------------------------------------
[8644](https://github.com/pytorch/pytorch/runs/6500788273?check_suite_focus=true#step:9:8645)
Traceback (most recent call last):
[8645](https://github.com/pytorch/pytorch/runs/6500788273?check_suite_focus=true#step:9:8646)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 1808, in wrapper
[8646](https://github.com/pytorch/pytorch/runs/6500788273?check_suite_focus=true#step:9:8647)
    method(*args, **kwargs)
[8647](https://github.com/pytorch/pytorch/runs/6500788273?check_suite_focus=true#step:9:8648)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 377, in instantiated_test
[8648](https://github.com/pytorch/pytorch/runs/6500788273?check_suite_focus=true#step:9:8649)
    result = test(self, **param_kwargs)
[8649](https://github.com/pytorch/pytorch/runs/6500788273?check_suite_focus=true#step:9:8650)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 786, in test_wrapper
[8650](https://github.com/pytorch/pytorch/runs/6500788273?check_suite_focus=true#step:9:8651)
    return test(*args, **kwargs)
[8651](https://github.com/pytorch/pytorch/runs/6500788273?check_suite_focus=true#step:9:8652)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 821, in dep_fn
[8652](https://github.com/pytorch/pytorch/runs/6500788273?check_suite_focus=true#step:9:8653)
    return fn(slf, *args, **kwargs)
[8653](https://github.com/pytorch/pytorch/runs/6500788273?check_suite_focus=true#step:9:8654)
  File "test_foreach.py", line 243, in test_binary_op_scalarlist_fastpath
[8654](https://github.com/pytorch/pytorch/runs/6500788273?check_suite_focus=true#step:9:8655)
    self._test_binary_op_scalarlist(device, dtype, op, N, scalarlist, True, disable_fastpath)
[8655](https://github.com/pytorch/pytorch/runs/6500788273?check_suite_focus=true#step:9:8656)
  File "test_foreach.py", line 216, in _test_binary_op_scalarlist
[8656](https://github.com/pytorch/pytorch/runs/6500788273?check_suite_focus=true#step:9:8657)
    self._binary_test(dtype, op, ref, inputs, is_fastpath, is_inplace=False)
[8657](https://github.com/pytorch/pytorch/runs/6500788273?check_suite_focus=true#step:9:8658)
  File "test_foreach.py", line 111, in _binary_test
[8658](https://github.com/pytorch/pytorch/runs/6500788273?check_suite_focus=true#step:9:8659)
    actual = op(inputs, self.is_cuda, is_fastpath)
[8659](https://github.com/pytorch/pytorch/runs/6500788273?check_suite_focus=true#step:9:8660)
  File "test_foreach.py", line 83, in __call__
[8660](https://github.com/pytorch/pytorch/runs/6500788273?check_suite_focus=true#step:9:8661)
    assert e.count == self.n_expected_cudaLaunchKernels
[8661](https://github.com/pytorch/pytorch/runs/6500788273?check_suite_focus=true#step:9:8662)
AssertionError
```

</details>

Local Run
```
pytest test/test_foreach.py -v -k test_binary_op_scalarlist_fastpath__foreach_div_cuda_float16
test/test_foreach.py::TestForeachCUDA::test_binary_op_scalarlist_fastpath__foreach_div_cuda_float16 PASSED                                                     [100%]
```

```
pytest test/test_foreach.py -v -k test_binary_op_scalarlist_fastpath__foreach_mul_cuda_float16
test/test_foreach.py::TestForeachCUDA::test_binary_op_scalarlist_fastpath__foreach_mul_cuda_float16 PASSED                                                     [100%]
```

File: test/test_type_promotion.py [hunk 10]
```diff
@@ -120,6 +124,87 @@ def test_promotion(dtype_float, dtype_double):
         other = torch.tensor(5.5, dtype=torch.double, device=device)
         self.assertEqual((a + other).dtype, torch.complex64)
 
+        def make_scalar(dtype):
```
Comment (mruberry) [inline test/test_type_promotion.py:127]: make_scalar_tensor

**Review by mruberry — APPROVED** [review]
LGTM! Nice work, @kshitij12345

**Comment by ngimel** [comment]
LIkely flaky

File: test/test_type_promotion.py [hunk 11]
```diff
@@ -120,6 +124,87 @@ def test_promotion(dtype_float, dtype_double):
...
+            return make_tensor((), dtype=dtype, device=device)
+
+        def make_1d_tensor(dtype):
+            return make_tensor((3,), dtype=dtype, device=device)
+
+        def complex_scalar_tensor_test(s, t, is_float_tensor):
```
Comment (ngimel) [inline test/test_type_promotion.py:133]: why is `is_float_tensor` an argument and not programmatically determined from `t.dtype`?

Comment (kshitij12345) [reply test/test_type_promotion.py:133]: Missed that. Thanks!

**Comment by kshitij12345** [comment]
Same tests from `test_foreach` failed again. Unable to reproduce them with local build (cuda-11.3)

Ref: https://github.com/pytorch/pytorch/runs/6512197886?check_suite_focus=true#step:9:9722

cc: @ngimel @anjali411 @crcrpar (original author)

**Comment by anjali411** [comment]
@kshitij12345 it's ok to disable the test and land this PR I think because the failure is not locally reproducible

**Pull request closed without merging**